Use sub action to build individual projects

yigit opened the pull request:
This CL refactors the workflow to use a sub action to build individual
projects. This helps cleanup the build and also have an easy way to
disable certain builds.

I've also added a new ci-config.json file to control which projects are
included / excluded along with a companion script to decide based on the
current branch.

For now, we have two setups, 1 for compose and 1 for the rest. Compose
branch only enables compose projects meanwhile the main branch enables
everything by default.

Bug: n/a
Test: CI

yigit committed:
Use sub action to build individual projects

This CL refactors the workflow to use a sub action to build individual
projects. This helps cleanup the build and also have an easy way to
disable certain builds.

I've also added a new ci-config.json file to control which projects are
included / excluded along with a companion script to decide based on the
current branch.

For now, we have two setups, 1 for compose and 1 for the rest. Compose
branch only enables compose projects meanwhile the main branch enables
everything by default.

Bug: n/a
Test: CI

yigit committed:
add a comment to refer people to ci-config

liutikas commented on .github/ci-control/ci-config.json:
```diff
@@ -0,0 +1,21 @@
+    "__comment": [
+        "Compose and Main are two configurations we have in CI.",
+        "Each can have include and exclude filters where these filters run a",
+        "  'contains' match on the project name. e.g. if the project name is compose-runtime",
```
Can we go with exact matches and validate it matches a real project? I am a bit worried this will lead to subtle errors.

yigit replied:
yea i flip-flopped here a lot on what to do. Another option is to allow matchers here (e.g. you can write `*compose*`).

yigit replied:
moving to exact string match for now.

liutikas commented on .github/actions/build-single-project/action.yml:
```diff
@@ -0,0 +1,106 @@
+    description: "password for gradle cache"
+  gradle-enterprise-access-key:
+    description: "access key for gradle enterprise"
+  gradle-flags:
```
Why is this one not required? We seem to always set it.

yigit replied:
they may not be available in the caller (actually, the secret is only available in post-submit builds)

liutikas commented on .github/ci-control/should_run_project.sh:
```diff
@@ -0,0 +1,79 @@
+    usage
+fi
+
+if [[ $BRANCH == *compose* ]]; then
```
Can we only match `compose-compiler-*`? I prefer we are as explicit as possible

yigit replied:
moved to "*compose-compiler*"

liutikas commented on .github/ci-control/should_run_project.sh:
```diff
@@ -0,0 +1,79 @@
+# Decides whether we should build a project in the current branch based on the ci-config.json file.
```
Not a blocker, but could we pick python or something else thats better than bash? :)

yigit replied:
i'm afraid to admit that i'm more familiar with bash than python :) And i don't know bash at all :) so maybe i should try to learn python and convert this :)

yigit replied:
moved to python. Should not review on full stomach.

liutikas commented on .github/workflows/presubmit.yml:
```diff
@@ -117,6 +116,7 @@ jobs:
       fail-fast: false
       matrix:
         os: [ubuntu-latest]
+        # if you would like to remove some projects temporarily, use ci-config.json instead.
```
Put pull path .github/ci-control/ci-config.json so people can find it.

liutikas reviewed (approved):
Could we add a README.md explaining how our CI is structured? It might be nice to make it easy to figure out how it works.

525411 commented:
Merge full

On Wed, Feb 2, 2022, 1:20 PM Aurimas, ***@***.***> wrote:

> ***@***.**** approved this pull request.
>
> Could we add a README.md explaining how our CI is structured? It might be
> nice to make it easy to figure out how it works.
> ------------------------------
>
> In .github/ci-control/ci-config.json
> <https://github.com/androidx/androidx/pull/322#discussion_r797273367>:
>
> > @@ -0,0 +1,21 @@
> +{
> +    "__comment": [
> +        "Compose and Main are two configurations we have in CI.",
> +        "Each can have include and exclude filters where these filters run a",
> +        "  'contains' match on the project name. e.g. if the project name is compose-runtime",
>
> Can we go with exact matches and validate it matches a real project? I am
> a bit worried this will lead to subtle errors.
> ------------------------------
>
> In .github/actions/build-single-project/action.yml
> <https://github.com/androidx/androidx/pull/322#discussion_r797275022>:
>
> > @@ -0,0 +1,106 @@
> +name: 'build-single-playground-project'
> +description: 'Builds one playground project based on inputs'
> +inputs:
> +  project:
> +    description: "Name of the project (used for artifacts)"
> +    required: true
> +  project-root:
> +    description: "Root folder of the project (used for building)"
> +    required: true
> +  gradle-cache-password:
> +    description: "password for gradle cache"
> +  gradle-enterprise-access-key:
> +    description: "access key for gradle enterprise"
> +  gradle-flags:
>
> Why is this one not required? We seem to always set it.
> ------------------------------
>
> In .github/ci-control/should_run_project.sh
> <https://github.com/androidx/androidx/pull/322#discussion_r797276238>:
>
> > +            usage
> +    esac
> +    shift
> +done
> +
> +if  [[ $PROJECT_NAME == "" ]]; then
> +    echo "Missing projectName argument"
> +    usage
> +fi
> +
> +if  [[ $BRANCH == "" ]]; then
> +    echo "Missing branch argument"
> +    usage
> +fi
> +
> +if [[ $BRANCH == *compose* ]]; then
>
> Can we only match compose-compiler-*? I prefer we are as explicit as
> possible
> ------------------------------
>
> In .github/ci-control/should_run_project.sh
> <https://github.com/androidx/androidx/pull/322#discussion_r797276799>:
>
> > @@ -0,0 +1,79 @@
> +# Decides whether we should build a project in the current branch based on the ci-config.json file.
>
> Not a blocker, but could we pick python or something else thats better
> than bash? :)
> ------------------------------
>
> In .github/workflows/presubmit.yml
> <https://github.com/androidx/androidx/pull/322#discussion_r797278108>:
>
> > @@ -117,6 +116,7 @@ jobs:
>        fail-fast: false
>        matrix:
>          os: [ubuntu-latest]
> +        # if you would like to remove some projects temporarily, use ci-config.json instead.
>
> Put pull path .github/ci-control/ci-config.json so people can find it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/androidx/androidx/pull/322#pullrequestreview-870094647>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXKJHJCXWBZJ746DRZT73FTUZC5J5ANCNFSM5NKOMACQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

yigit committed:
move to python, temporary disable to test on CI

yigit committed:
remove bash script

yigit committed:
format python script

yigit committed:
remove testing exclusions, include compose-runtime in compose branches

dlam commented on .github/ci-control/should_run_project.py:
```diff
@@ -0,0 +1,65 @@
+def usage(exitCode):
+    print("""
+        Usage:
+        ./should_run_project.sh --project <project-name> --branch <current branch name>\n\
```
```suggestion
        ./should_run_project.py --project <project-name> --branch <current branch name>\n\
```

yigit committed:
fix usage text

Co-authored-by: Dustin Lam <dustinlam@google.com>

yigit committed:
Move artifact steps back into the main action

They belong to the sub action but it doesn't support
continue-on0error there. Furthermore, this upload
infra is not very stable on github and we don't want
its random failures. Hence, I'm moving them back into
the main action until we can be sure thaat they won't fail
randomly (or better, have continue-on-error support
in sub actions)

525411 commented:
Idonate all up to u.ok thanks...

On Fri, Feb 4, 2022, 4:20 AM Yigit Boyar ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In .github/ci-control/ci-config.json
> <https://github.com/androidx/androidx/pull/322#discussion_r798937041>:
>
> > @@ -0,0 +1,21 @@
> +{
> +    "__comment": [
> +        "Compose and Main are two configurations we have in CI.",
> +        "Each can have include and exclude filters where these filters run a",
> +        "  'contains' match on the project name. e.g. if the project name is compose-runtime",
>
> moving to exact string match for now.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/androidx/androidx/pull/322#discussion_r798937041>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXKJHJF34AUJNLX4C5JDUKLUZLPQ7ANCNFSM5NKOMACQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

yigit committed:
Add a readme for ci-control

yigit commented:
added a README for the ci-config folder.

yigit committed:
Fix lint errors
